### PR TITLE
Fix issues with AsyncLocal activity scope stack implementation

### PR DESCRIPTION
**Pull request by masonlamy**
Using ```AsyncLocal<Stack<ActivityScope>>```  creates race conditions and allows unexpected changes to occur to the activity scope stack when there are multiple async children running.  Implementing ```AsyncLocal<ImmutableStack<ActivityScope>>``` enforces copying and prevents these issues.

One particular issue it addresses is an exception we had been seeing of the form:
```
Stack empty. System.InvalidOperationException: Stack empty.\r\n   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)\r\n   at System.Collections.Generic.Stack`1.Pop()\r\n   at CorrelatorSharp.ActivityTracker.End(ActivityScope scope)\r\n   at CorrelatorSharp.ActivityScope.Dispose()\r\n
```

It also addresses the issue where if you spawn two portions of async work, the second portion of async work would be inherit the first async as it's parent even if actually independent.

Added relevant test for this.

### CODE REVIEW

File: CorrelatorSharp/ActivityTracker.cs
Comment (mikegoatly) [inline CorrelatorSharp/ActivityTracker.cs]: Not sure why this loop would be needed - surely there should only ever be 1 push for 1 pop; popping multiple times on End feels like plastering over something that was incorrect before?

**Comment by mikegoatly** [comment]
Looks good to me. Whoever creates the nuget package would need to remember to add a reference to the ImmutableCollections package.

File: CorrelatorSharp/Properties/AssemblyInfo.cs
Comment (ivanz) [inline CorrelatorSharp/Properties/AssemblyInfo.cs:35]: Packave versioning is controlled in `appveyor.yml` and not here. Please update the yml file :)

**Review by ivanz — CHANGES_REQUESTED** [review]
deleted

**Comment by ivanz** [comment]
I hadn't considered this scenario before and I had assumed that Task.Run will not flow the async context :( Am I thinking right that we are still exposed to at least the following race condition?
1. Root task starts. Stack; `root`
2. Child A spawned. Stack: `root -> A`
   3 Child B spawned (parallel). Stack replaced: `root -> B`
3. Child A spawns task C: `root->B->C` <-- wrong chain

**Comment by mikegoatly** [comment]
The data in the async local is copied whenever it is modified in an async call context, so as long as you have an immutable type stored in it, you won't run into the problem that you're describing*. I've [created a gist](https://gist.github.com/mikegoatly/e1a6db7e875dc3c547b86271aab90746) to demonstrate this behaviour.

\* Stephen Cleary [has already described this](http://blog.stephencleary.com/2013/04/implicit-async-context-asynclocal.html) much more eloquently than I can, so I'll let him explain the mechanics in full!

**Comment by masonlamy** [comment]
Hey Ivan, good to hear back from you.  Hope you're well.  Thanks @mikegoatly for following up on that - as Mike says the race condition highlighted won't be an issue with this fix.   I'll update the appveyor.yml file.

**Comment by ivanz** [comment]
@mikegoatly Great bit of info to know - thanks for chipping in. 

@masoapps The final thing left is to add an explicit version range for the immutable collections dependency in the `.nuspec` - `[1.0,2.0)` - otherwise the package ends up with a fixed dep on specific version (1.0.34). I will merge it once you do that.

**Comment by masonlamy** [comment]
Thanks @ivanz, I've updated the branch with the version range specified in the nuspec dependencies, let me know if there's any other changes needed before merging.

File: CorrelatorSharp/packages.config
Comment (ivanz) [inline CorrelatorSharp/packages.config]: By removing this a fresh checkout of the repo won't pull in the dependency on build unless I am missing something?

**Review by ivanz — CHANGES_REQUESTED** [review]

File: CorrelatorSharp/packages.config
Comment (masonlamy) [reply CorrelatorSharp/packages.config]: I'm not too familiar with nuspec/nuget configurations.  I may be mistaken but I believed when there is no packages.config it should check the nuspec dependencies and pull them down if required? I performed a clean checkout and build locally and it seemed to be okay but maybe there was some unexpected caching somewhere... I'll add packages.config back in and make immutable collections a developmentDependency.

Comment (ivanz) [reply CorrelatorSharp/packages.config]: .nuspec is used to create the .nupkg and packages.config is used at build/nuget restore time to download the required dependencies from nuget.org. When you install a package in a project in Visual Studio - it adds it to the packages.config. So, if I check out your branch - I won't be able to build it now, because it won't download and install the Immutable collections package.

Comment (masonlamy) [reply CorrelatorSharp/packages.config]: Thanks for clarifying Ivan, I reproduced that failure after removing the package from my local user .nuget cache.  Have added the packages.config back with my latest commit

**Review by ivanz — APPROVED** [review]

**Comment by ivanz** [comment]
Thanks! Merged (with one minor change not to change the assembly version, because that leads to headaches with warnings when compiling non-strongly signed assemblies). New version should appear on NuGet.org automatically shortly

**Comment by masonlamy** [comment]
That's great!  Cheers Ivan

**Pull request merged**
